src/farming/library.cairo

```diff
@@ -0,0 +1,1039 @@
+            assert_uint256_le(value, project_value);
+        }
+
+        // [Compute] Otherwise returns the absorption corresponding to the ratio
+        // Keep quotient_low only since value <= project_value
+        let (quotient_low, quotient_high, _) = uint256_mul_div_mod(value, total_absorption_u256, project_value);
+
+        // [Check] No underflow
+        let zero = Uint256(low=0, high=0);
+        with_attr error_message("CarbonableFarming: Unexpected underflow") {
+            assert_uint256_eq(quotient_high, zero);
+        }
+
+        let (absorption) = _uint_to_felt(quotient_low);
+        return (absorption=absorption);
+    }
+
+    func _compute_total_sale{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        sale: felt
+    ) {
+        alloc_locals;
+
+        // [Compute] Total deposited value and registered time
+        let (value) = CarbonableFarming_total_registered_value_.read();
+        let (time) = CarbonableFarming_total_registered_time_.read();
+
+        // [Compute] Total computed sale
+        return _compute_sale(value=value, time=time);
+    }
+
+    func _compute_user_sale{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        address: felt
+    ) -> (sale: felt) {
+        alloc_locals;
+
+        // [Compute] User deposited value and registered time
+        let (value) = CarbonableFarming_registered_value_.read(address);
+        let (time) = CarbonableFarming_registered_time_.read(address);
+
+        // [Compute] User computed sale
+        return _compute_sale(value=value, time=time);
+    }
+
+    func _compute_sale{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        value: Uint256, time: felt
+    ) -> (sale: felt) {
+        alloc_locals;
+
+        // [Check] Deposited value not null
+        let zero = Uint256(low=0, high=0);
+        let (is_zero) = uint256_eq(value, zero);
+        if (is_zero == TRUE) {
+            return (sale=0);
+        }
+
+        // [Compute] Project value is not zero
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (project_value) = ICarbonableProject.getProjectValue(
+            contract_address=contract_address, slot=slot
+        );
+        let (is_zero) = uint256_eq(project_value, zero);
+        if (is_zero == TRUE) {
+            return (sale=0);
+        }
+
+        // [Compute] Interpolated sales
+        let (current_time) = get_block_timestamp();
+        let (len, times, _, _, cumsales) = CarbonableFarming._compute_cumsales();
+
+        // [Check] Times and prices are set
+        if (len == 0) {
+            return (sale=0);
+        }
+
+        // [Compute] Total sale
+        let initial_cumsale = Math.interpolate(
+            x=time, len=len, xs=times, ys=cumsales, interpolation=LINEAR, extrapolation=CONSTANT
```

Comment:
Format , all other `Math.interpolate` variables are in column

Reply:
I think cairo-format puts it like this automatically but there might be a command-line argument to tell it not to.

Reply:
I forgot to run code formatter after latest updates